lib/boombox.ex

```diff
@@ -125,41 +125,97 @@ defmodule Boombox do
           input: input(),
           output: output()
         ) :: :ok | Enumerable.t()
-  @endpoint_opts [:input, :output]
   def run(stream \\ nil, opts) do
-    opts =
-      opts
-      |> Keyword.validate!(@endpoint_opts)
-      |> Map.new(fn {key, value} -> {key, parse_endpoint_opt!(key, value)} end)
+    opts = validate_opts!(stream, opts)
 
-    :ok = maybe_log_transcoding_related_warning(opts)
+    case opts do
+      %{input: {:stream, _stream_opts}} ->
+        procs = start_pipeline(opts)
+        source = await_source_ready()
+        consume_stream(stream, source, procs)
+
+      %{output: {:stream, _stream_opts}} ->
+        procs = start_pipeline(opts)
+        sink = await_sink_ready()
+        produce_stream(sink, procs)
 
-    if key = Enum.find(@endpoint_opts, fn k -> not is_map_key(opts, k) end) do
-      raise "#{key} is not provided"
+      opts ->
+        opts
+        |> start_pipeline()
+        |> await_pipeline()
     end
+  end
 
-    case opts do
-      %{input: {:stream, _in_stream_opts}, output: {:stream, _out_stream_opts}} ->
-        raise ArgumentError, ":stream on both input and output is not supported"
+  @doc """
+  Asynchronous version of run/2
+  Doesn't block the calling process until the termination of the pipeline.
+  It returns a `Task` that can be awaited later.
+  If the output is a stream the behaviour is idential to run/2
+  """
+  @spec async(Enumerable.t() | nil,
+          input: input(),
+          output: output()
+        ) :: Task.t() | Enumerable.t()
+  def async(stream \\ nil, opts) do
+    opts = validate_opts!(stream, opts)
 
+    case opts do
       %{input: {:stream, _stream_opts}} ->
-        unless Enumerable.impl_for(stream) do
-          raise ArgumentError,
-                "Expected Enumerable.t() to be passed as the first argument, got #{inspect(stream)}"
-        end
+        procs = start_pipeline(opts)
+        source = await_source_ready()
 
-        consume_stream(stream, opts)
+        Task.async(fn ->
+          Process.monitor(procs.supervisor)
+          consume_stream(stream, source, procs)
+        end)
 
       %{output: {:stream, _stream_opts}} ->
-        produce_stream(opts)
+        procs = start_pipeline(opts)
+        sink = await_sink_ready()
+        produce_stream(sink, procs)
 
       opts ->
+        procs = start_pipeline(opts)
```

Comment:
With this `async()` implementation we will wait until the `source` enters `playing` playback (which means that the pipeline's playback is already `:playing`) for input being a stream, but we won't wait for pipeline entering `:playing` playback for non-stream inputs and outputs. Perhaps it would be better to add some guarantee that when `async()` returns, the pipeline's playback is already `:playing`?